<a href="https://colab.research.google.com/github/NV1511/Hotel-Booking-Data-Analysis/blob/main/Capstone_Project_Hotel_Booking_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **1. Problem Statement**

<b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

<b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

<b> Explore and analyze the data to discover important factors that govern the bookings. </b>

### Steps performed in this EDA *Projests*

Handling this dataset with the fundamental steps to unvail the factors :

* Imports Libraries, Dataset and Other modules
* Analysing Data : Reading & Inspection of Data.
    *   Data Structure (Head , Tail or Body)
    *   Data information (basic schema)
    *   Arithmatic measurement (data discriptions)
* Data Preparation & Processing
    *   Identifying Different Variable
    *   Sortlisting Dependent Variables 
    *   Handling Null Values
    *   Adding, combining &renaming of Column if required.
* Analysis & Visualization
    *   Identifying relation of dependent variables with Problemset.
    *   Visualization using differen plots in relation with question.
* Conclusion : Answer to the Questions


# **2. Import required libraries and get data for analysis**

In [3]:
# Import Iibraries
import pandas as pd  
import numpy as np
from datetime import date
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
path='/content/drive/MyDrive/AlmaBetter /Capstone Project/Capstone Project - 1/Hotel Bookings.csv'
df=pd.read_csv(path)

# **3. Review the data and analyse all columns**

In [6]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [8]:
df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

# **4. Data Preparation & Processing**

## Type Of Varaiables
### The dataset have **<font color='green'>32 variables (Continuous and Categorical)</font>**
Observation:

All the data type are in required format only, and no need to convert or change the datatype,Some Visible Null Values in certain column.

Total Rows = 119390, Column = 32

## Sortlisting of Dependent Variables
Well after brainstorming we conclude the following set

##**Categorical**  (20 Nos) :

<font color='red'>[ hotel, is_canceled, arrival_date_year, arrival_date_month, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, reserved_room_type, assinged_room_type, deposit_type, agent, company, customer_type, required_car_parking_spaces, total_of_special_requests, reservation_status ]</font>

Out of this 20 variable we found 10 variable which are dependent with the problem set and expecting high or medium relevence to th problem.

* hotel, 
* arrival_date_month, 
* children, 
* babies, 
* country, 
* market_segment, 
* distribution_channel, 
* is_repeated_guest,
* company,
* customer_type

## **Continuous** (12 Nos) 

<font color='red'>[lead_time, arrival_date_week_number, arrival_date_day_of_month, staysinweekend_nights, stays_in_week_nights, adults, previous_cancellations, previous_bookings_not_canceled, booking_changes, days_in_waiting_list, sdr, reservation_status_date ]</font>

Out of 12 Nos of Continuous variable we sortlisted to 6 variable which are dependent with the problem set and expecting high or medium relevence to th problem.

* lead_time,  
* arrival_date_day_of_month, 
* stays_in_weekend_nights, 
* stays_in_week_nights, 
* adults,  
* adr, 

## **Handling Null Values**

In [9]:
Total = df.isnull().sum().sort_values(ascending=False)
Percentage = (df.isnull().sum()/df.isnull().count()*100).sort_values(ascending=False)
Missingdata = pd.concat([Total,Percentage], axis=1, keys=['Total', 'Percentage'])
Missingdata.head(10)

Total  Percentage
company                 112593   94.306893
agent                    16340   13.686238
country                    488    0.408744
children                     4    0.003350
reserved_room_type           0    0.000000
assigned_room_type           0    0.000000
booking_changes              0    0.000000
deposit_type                 0    0.000000
hotel                        0    0.000000
previous_cancellations       0    0.000000

variable Company has 94.3% of missing data and agent has 13.69%,

we will Simply drop the high missing value column, thing is Company is in dependent Variable list but with 94% missing data variable has no use.

for Country the is 488 rows with null values comparing 488 to total numbers 0.5% so we will remove those rows only, alsochildren has four missing values we will fill with mean values, rest all are fine.

In [10]:
df = df.drop(['company','agent'], axis=1)
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118898 entries, 0 to 119389
Data columns (total 30 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           118898 non-null  object 
 1   is_canceled                     118898 non-null  int64  
 2   lead_time                       118898 non-null  int64  
 3   arrival_date_year               118898 non-null  int64  
 4   arrival_date_month              118898 non-null  object 
 5   arrival_date_week_number        118898 non-null  int64  
 6   arrival_date_day_of_month       118898 non-null  int64  
 7   stays_in_weekend_nights         118898 non-null  int64  
 8   stays_in_week_nights            118898 non-null  int64  
 9   adults                          118898 non-null  int64  
 10  children                        118898 non-null  float64
 11  babies                          118898 non-null  int64  
 12  meal            

We have now 30 column with zero null values.

we will check other possiblilities here one of the threat that sum of all customer i.e Adult + children + babies should be greater than zero. lets see

total 170 rows are there having sum of adults, children and babies is zero will remove these rows

In [11]:
df = df.drop(df[(df.adults+df.babies+df.children)==0].index)

#lets confirm
df[(df.adults+df.babies+df.children)==0].shape

(0, 30)

## **Process and add required columns**

We need to add Date of arrival by joining the mothn,year and date colun and after that we'll get month-year colunm which is require to do month wise booking analysis

In [12]:
#Add a month Number column
month = {'January':1,'February':2, 'March':3, 'April':4, 'May':5, 'June':6, 'July':7, 'August':8, 'September':9, 'October':10, 'November':11, 'December':12}

df['month_no']=0
for index,row in df.iterrows():
  df.loc[index,'month_no']=month[row['arrival_date_month']]

In [14]:
#Add Arival Date Column
df['dateInt']=df['arrival_date_year'].astype(str) + df['month_no'].astype(str).str.zfill(2)+ df['arrival_date_day_of_month'].astype(str).str.zfill(2)
df['Date'] = pd.to_datetime(df['dateInt'], format='%Y%m%d')

In [15]:
#Add Month-year Column
df['month_year'] = pd.to_datetime(df['Date']).dt.to_period('M')
